### DISCUSSION TITLE
LFPykit and Arbor

### DISCUSSION THREAD

**Discussion by brenthuisman** [arbor-sim/arbor#1806] [General] [answered]
Yesterday I spoke with @davbass about Arbor and he asked how Arbor can be used to estimate local field potentials. I forgot that we already have an example demonstrating combining LFPykit and Arbor, see [python/example/single_cell_extracellular_potentials.py](https://github.com/arbor-sim/arbor/blob/dc371e5d15fa6acc4b778c12df1f5c6e4adaadc2/python/example/single_cell_extracellular_potentials.py). @espenhgn kindly provided the example, so I thought that he might comment: is there anything more brewing on your end, or does this example show the integration between Arbor and LFpykit is 'complete'?

**Comment by espenhgn** [comment, marked as answer]
Hi Brent, 
I didn't have any immediate plans to change this example, although it could be improved somewhat:
- Providing a modified `lfpykit.CellGeometry` class for dealing with the mapping of geometries of segments to the corresponding CVs, taking the `place_pwlin` object and location metadata as inputs. I don't think the user should have to deal with this each time in order to obtain this matrix `M` that is finally multiplied with the transmembrane currents from the `cable_probe_total_current_cell` instance. 
- I don't know how long simulation times and typical number of CVs @davbass expects, but recording transmembrane currents for the entire simulation duration could easily result in consuming all available memory. I'll assume that the number of extracellular measurement locations are typically less than the number of CVs (in contrast to this example), hence I would like a solution allowing for:
-- precomputing this mapping of transmembrane currents to extracellular potentials (or any other linearly dependent measure)
-- push the corresponding multiplication to the internals of Arbor and rather record the output directly during the course of the simulation. 

Right now I believe this latter idea is not possible, as the CV data needed is not exposed before the neuron model/simulation is actually instantiated. 

@davbass; if you have any questions about LFPykit don't hesitate to contact me (via https://github.com/LFPy/LFPykit/issues for instance)

**Comment by brenthuisman** [comment]
Thanks @espenhgn ! I think this example should be promoted to a tutorial, because people have asked about this before and it would be good be able to show a working demo. @espenhgn are you planning to modify `CellGeometry` in the short term? From what I understood, @davbass would be starting a study in the near future, so I hope this is helpful. Indeed, don't hesitate to follow up, @davbass !

**Reply by espenhgn** [reply]
Hi. A tutorial sounds like a good idea. By that, it looks like I would only have to add the tutorial file(s) with some additional info here https://github.com/arbor-sim/arbor/tree/master/doc/tutorial in a PR, correct? As for the LFPykit classes I think in this first iteration it can be incorporated in the example file directly.

**Reply by brenthuisman** [reply]
Exactly. The only technical requirement is that we use the `literalinclude` directive to inline relevant parts of the example code, such that they can't diverge. Do you have time to write the tutorial?

**Reply by espenhgn** [reply]
Yes, I think I can find the time to create such a tutorial in the next 2-3 weeks.

**Reply by brenthuisman** [reply]
Great, looking forward!

**Comment by brenthuisman** [comment]
In the meantime, the tutorial has been added: https://docs.arbor-sim.org/en/latest/tutorial/tutorial_lfpykit.html
